File: udon_decompiler/analysis/transform/passes/high_level_loop_statement_transform.py
from __future__ import annotations

from typing import Optional, cast

from udon_decompiler.analysis.ir.nodes import (
    IRBlock,
    IRBlockContainer,
    IRContainerKind,
    IRFunction,
    IRHighLevelDoWhile,
    IRHighLevelSwitch,
    IRHighLevelWhile,
    IRIf,
    IRJump,
    IRLeave,
    IRStatement,
)
from udon_decompiler.analysis.transform.pass_base import (
    ITransform,
    TransformContext,
)


class HighLevelLoopStatementTransform(ITransform):
    """
    Lift low-level loop containers into explicit high-level loop statements.

    - `ContainerKind.Loop`      -> `while (true)`
    - `ContainerKind.While`     -> `while (condition)`
    - `ContainerKind.DoWhile`   -> `do { ... } while (condition)`
    """

    def run(self, function: IRFunction, context: TransformContext) -> None:
        self._context = context
        self._rewrite_container(function.body)

    def _rewrite_container(self, container: IRBlockContainer) -> None:
        for block in list(container.blocks):
            block.statements = [
                self._rewrite_statement(statement) for statement in block.statements
            ]

    def _rewrite_statement(self, statement: IRStatement) -> IRStatement:
        if isinstance(statement, IRIf):
            statement.true_statement = self._rewrite_statement(statement.true_statement)
            if statement.false_statement is not None:
                statement.false_statement = self._rewrite_statement(
                    statement.false_statement
                )
            return statement

        if isinstance(statement, IRBlock):
            statement.statements = [
                self._rewrite_statement(nested) for nested in statement.statements
            ]
            return statement

        if isinstance(statement, IRHighLevelSwitch):
            for section in statement.sections:
                section.body = cast(IRBlock, self._rewrite_statement(section.body))
            return statement

        if isinstance(statement, IRBlockContainer):
            self._rewrite_container(statement)
            return self._lift_container(statement)

        return statement

    def _lift_container(self, container: IRBlockContainer) -> IRStatement:
        if container.kind == IRContainerKind.LOOP:
            return self._lift_infinite_loop(container)

        if container.kind == IRContainerKind.WHILE:
            while_stmt = self._lift_while(container)
            if while_stmt is not None:
                return while_stmt
            return container

        if container.kind == IRContainerKind.DO_WHILE:
            do_while_stmt = self._lift_do_while(container)
            if do_while_stmt is not None:
                return do_while_stmt
            return container

        return container

    def _lift_infinite_loop(
        self,
        container: IRBlockContainer,
    ) -> IRStatement:
        entry = container.entry_block
        if entry is None:
            return container

        body = IRBlockContainer(
            blocks=list(container.blocks),
            kind=IRContainerKind.BLOCK,
        )
        self._remove_trailing_continue_jump(body, entry)
        return IRHighLevelWhile(
            condition=None,
            body=body,
            continue_target=entry,
            break_target=container,
        )

    def _lift_while(
        self,
        container: IRBlockContainer,
    ) -> Optional[IRStatement]:
        entry = container.entry_block
        if entry is None:
            return None
        if len(entry.statements) != 1:
            return None

        if_inst = entry.statements[0]
        if not isinstance(if_inst, IRIf):
            return None
        if if_inst.false_statement is None:
            return None
        if not self._is_leave_to_container(if_inst.false_statement, container):
            return None

        inline_loop_body_block: Optional[IRBlock] = None
        loop_body_target: Optional[IRBlock] = None

        if isinstance(if_inst.true_statement, IRJump):
            loop_body_target = if_inst.true_statement.target
            if loop_body_target not in container.blocks:
                return None
        elif isinstance(if_inst.true_statement, IRBlock):
            inline_loop_body_block = IRBlock(
                statements=list(if_inst.true_statement.statements),
                start_address=self._next_synthetic_block_address(),
            )
        else:
            return None

        # The condition block becomes a pure continue label anchor.
        entry.statements = []

        body_blocks: list[IRBlock] = []
        if inline_loop_body_block is not None:
            body_blocks.append(inline_loop_body_block)
        elif loop_body_target is not None:
            body_blocks.append(loop_body_target)

        for block in container.blocks[1:]:
            if loop_body_target is not None and block is loop_body_target:
                continue
            body_blocks.append(block)

        # Keep a physical anchor block for gotos that cannot be rendered as continue.
        body_blocks.append(entry)

        body_container = IRBlockContainer(
            blocks=body_blocks,
            kind=IRContainerKind.BLOCK,
        )
        self._remove_trailing_continue_jump(body_container, entry)

        return IRHighLevelWhile(
            condition=if_inst.condition,
            body=body_container,
            continue_target=entry,
            break_target=container,
        )

    def _lift_do_while(
        self,
        container: IRBlockContainer,
    ) -> Optional[IRStatement]:
        entry = container.entry_block
        if entry is None or not container.blocks:
            return None

        condition_block = container.blocks[-1]
        if not condition_block.statements:
            return None

        if_inst = condition_block.statements[-1]
        if not isinstance(if_inst, IRIf):
            return None
        if if_inst.false_statement is None:
            return None
        if not self._is_leave_to_container(if_inst.false_statement, container):
            return None
        if not self._is_jump_to_entry(if_inst.true_statement, entry):
            return None

        body_blocks = list(container.blocks[:-1])
        condition_prefix = condition_block.statements[:-1]

        if condition_prefix:
            body_blocks.append(
                IRBlock(
                    statements=list(condition_prefix),
                    start_address=self._next_synthetic_block_address(),
                )
            )

        # The original condition block becomes a pure continue label anchor.
        condition_block.statements = []
        body_blocks.append(condition_block)

        body_container = IRBlockContainer(
            blocks=body_blocks,
            kind=IRContainerKind.BLOCK,
        )
        self._remove_trailing_continue_jump(body_container, condition_block)

        return IRHighLevelDoWhile(
            condition=if_inst.condition,
            body=body_container,
            continue_target=condition_block,
            break_target=container,
        )

    @staticmethod
    def _is_leave_to_container(
        statement: IRStatement,
        container: IRBlockContainer,
    ) -> bool:
        return (
            isinstance(statement, IRLeave) and statement.target_container is container
        )

    @staticmethod
    def _is_jump_to_entry(
        statement: IRStatement,
        entry: IRBlock,
    ) -> bool:
        if isinstance(statement, IRJump):
            return statement.target is entry
        if isinstance(statement, IRBlock) and len(statement.statements) == 1:
            nested = statement.statements[0]
            return isinstance(nested, IRJump) and nested.target is entry
        return False

    def _next_synthetic_block_address(self) -> int:
        key = "_synthetic_block_addr"
        current = self._context.metadata.get(key)
        if not isinstance(current, int):
            current = -1
        self._context.metadata[key] = current - 1
        return current

    @staticmethod
    def _remove_trailing_continue_jump(
        body: IRBlockContainer,
        continue_target: IRBlock,
    ) -> None:
        for block in reversed(body.blocks):
            if not block.statements:
                continue
            last = block.statements[-1]
            if isinstance(last, IRJump) and last.target is continue_target:
                block.statements.pop()
            break

File: src/decompiler/transform/passes/high_level_loop_statement_transform.rs
use crate::decompiler::Result;
use crate::decompiler::ir::{
    IrBlock, IrBlockContainer, IrContainerKind, IrFunction, IrHighLevelDoWhile, IrHighLevelWhile,
    IrJump, IrLeave, IrStatement,
};
use crate::decompiler::transform::pass_base::{ITransform, TransformContext};

pub struct HighLevelLoopStatementTransform;

impl ITransform for HighLevelLoopStatementTransform {

    fn run(&self, function: &mut IrFunction, context: &mut TransformContext<'_, '_>) -> Result<()> {
        let mut state = LoopStatementState::from_context(function, context);
        rewrite_container(&mut function.body, &mut state);
        state.commit(context);
        Ok(())
    }
}

#[derive(Debug, Clone)]
struct LoopStatementState {
    next_container_id: u32,
    synthetic_block_address: i64,
}

impl LoopStatementState {
    fn from_context(function: &IrFunction, context: &TransformContext<'_, '_>) -> Self {
        let current = context
            .program_context
            .metadata
            .get("_synthetic_block_addr")
            .copied()
            .unwrap_or(-1);
        Self {
            next_container_id: max_container_id(&function.body).saturating_add(1),
            synthetic_block_address: current,
        }
    }

    fn next_container_id(&mut self) -> u32 {
        let current = self.next_container_id;
        self.next_container_id = self.next_container_id.saturating_add(1);
        current
    }

    fn next_synthetic_block_address(&mut self) -> u32 {
        let current = self.synthetic_block_address;
        self.synthetic_block_address -= 1;
        (current as i32) as u32
    }

    fn commit(self, context: &mut TransformContext<'_, '_>) {
        context.program_context.metadata.insert(
            "_synthetic_block_addr".to_string(),
            self.synthetic_block_address,
        );
    }
}

fn rewrite_container(container: &mut IrBlockContainer, state: &mut LoopStatementState) {
    for block in &mut container.blocks {
        let mut rewritten = Vec::<IrStatement>::with_capacity(block.statements.len());
        for statement in block.statements.drain(..) {
            rewritten.push(rewrite_statement(statement, state));
        }
        block.statements = rewritten;
    }
}

fn rewrite_statement(statement: IrStatement, state: &mut LoopStatementState) -> IrStatement {
    match statement {
        IrStatement::If(mut if_stmt) => {
            if_stmt.true_statement = Box::new(rewrite_statement(*if_stmt.true_statement, state));
            if let Some(false_statement) = if_stmt.false_statement.take() {
                if_stmt.false_statement =
                    Some(Box::new(rewrite_statement(*false_statement, state)));
            }
            IrStatement::If(if_stmt)
        }
        IrStatement::Block(mut block) => {
            let mut rewritten = Vec::<IrStatement>::with_capacity(block.statements.len());
            for nested in block.statements.drain(..) {
                rewritten.push(rewrite_statement(nested, state));
            }
            block.statements = rewritten;
            IrStatement::Block(block)
        }
        IrStatement::HighLevelSwitch(mut switch_stmt) => {
            for section in &mut switch_stmt.sections {
                let rewritten = rewrite_statement(IrStatement::Block(section.body.clone()), state);
                if let IrStatement::Block(block) = rewritten {
                    section.body = block;
                }
            }
            IrStatement::HighLevelSwitch(switch_stmt)
        }
        IrStatement::BlockContainer(mut container) => {
            rewrite_container(&mut container, state);
            lift_container(container, state)
        }
        _ => statement,
    }
}

fn lift_container(container: IrBlockContainer, state: &mut LoopStatementState) -> IrStatement {
    match container.kind {
        IrContainerKind::Loop => lift_infinite_loop(container, state),
        IrContainerKind::While => lift_while(container, state),
        IrContainerKind::DoWhile => lift_do_while(container, state),
        _ => IrStatement::BlockContainer(container),
    }
}

fn lift_infinite_loop(
    container: IrBlockContainer,
    state: &mut LoopStatementState,
) -> IrStatement {
    let Some(entry) = container.entry_block().cloned() else {
        return IrStatement::BlockContainer(container);
    };

    let mut body = IrBlockContainer {
        id: state.next_container_id(),
        blocks: container.blocks.clone(),
        kind: IrContainerKind::Block,
        should_emit_exit_label: container.should_emit_exit_label,
    };
    remove_trailing_continue_jump(&mut body, entry.start_address);

    IrStatement::HighLevelWhile(IrHighLevelWhile {
        condition: None,
        body,
        continue_target: entry.start_address,
        break_target: container.id,
    })
}

fn lift_while(container: IrBlockContainer, state: &mut LoopStatementState) -> IrStatement {
    let Some(entry) = container.entry_block().cloned() else {
        return IrStatement::BlockContainer(container);
    };
    if entry.statements.len() != 1 {
        return IrStatement::BlockContainer(container);
    }

    let IrStatement::If(if_stmt) = &entry.statements[0] else {
        return IrStatement::BlockContainer(container);
    };
    let Some(false_statement) = if_stmt.false_statement.as_ref() else {
        return IrStatement::BlockContainer(container);
    };
    if !is_leave_to_container(false_statement.as_ref(), container.id) {
        return IrStatement::BlockContainer(container);
    }

    let mut inline_loop_body_block: Option<IrBlock> = None;
    let mut loop_body_target: Option<IrBlock> = None;

    if let Some(target_addr) = as_jump_target_addr(if_stmt.true_statement.as_ref()) {
        if let Some(target_block) = container
            .blocks
            .iter()
            .find(|block| block.start_address == target_addr)
            .cloned()
        {
            loop_body_target = Some(target_block);
        } else {
            return IrStatement::BlockContainer(container);
        }
    } else if let IrStatement::Block(inline_block) = if_stmt.true_statement.as_ref() {
        inline_loop_body_block = Some(IrBlock {
            statements: inline_block.statements.clone(),
            start_address: state.next_synthetic_block_address(),
            should_emit_label: false,
        });
    } else {
        return IrStatement::BlockContainer(container);
    }

    if loop_body_target
        .as_ref()
        .is_some_and(|target| target.start_address == entry.start_address)
    {
        return IrStatement::BlockContainer(container);
    }

    let mut entry_anchor = entry.clone();
    entry_anchor.statements.clear();

    let mut body_blocks = Vec::<IrBlock>::new();
    if let Some(inline_loop_body_block) = inline_loop_body_block {
        body_blocks.push(inline_loop_body_block);
    } else if let Some(loop_body_target) = loop_body_target.as_ref() {
        body_blocks.push(loop_body_target.clone());
    }

    for block in container.blocks.iter().skip(1) {
        if loop_body_target
            .as_ref()
            .is_some_and(|target| target.start_address == block.start_address)
        {
            continue;
        }
        body_blocks.push(block.clone());
    }

    body_blocks.push(entry_anchor.clone());

    let mut body_container = IrBlockContainer {
        id: state.next_container_id(),
        blocks: body_blocks,
        kind: IrContainerKind::Block,
        should_emit_exit_label: container.should_emit_exit_label,
    };
    remove_trailing_continue_jump(&mut body_container, entry_anchor.start_address);

    IrStatement::HighLevelWhile(IrHighLevelWhile {
        condition: Some(if_stmt.condition.clone()),
        body: body_container,
        continue_target: entry_anchor.start_address,
        break_target: container.id,
    })
}

fn lift_do_while(container: IrBlockContainer, state: &mut LoopStatementState) -> IrStatement {
    let Some(entry) = container.entry_block().cloned() else {
        return IrStatement::BlockContainer(container);
    };
    if container.blocks.is_empty() {
        return IrStatement::BlockContainer(container);
    }

    let Some(condition_block) = container.blocks.last().cloned() else {
        return IrStatement::BlockContainer(container);
    };
    if condition_block.statements.is_empty() {
        return IrStatement::BlockContainer(container);
    }

    let Some(IrStatement::If(if_stmt)) = condition_block.statements.last() else {
        return IrStatement::BlockContainer(container);
    };
    let Some(false_statement) = if_stmt.false_statement.as_ref() else {
        return IrStatement::BlockContainer(container);
    };
    if !is_leave_to_container(false_statement.as_ref(), container.id) {
        return IrStatement::BlockContainer(container);
    }
    if !is_jump_to_entry(if_stmt.true_statement.as_ref(), entry.start_address) {
        return IrStatement::BlockContainer(container);
    }

    let mut body_blocks = container.blocks[..container.blocks.len() - 1].to_vec();
    let condition_prefix = condition_block
        .statements
        .iter()
        .take(condition_block.statements.len().saturating_sub(1))
        .cloned()
        .collect::<Vec<_>>();

    if !condition_prefix.is_empty() {
        body_blocks.push(IrBlock {
            statements: condition_prefix,
            start_address: state.next_synthetic_block_address(),
            should_emit_label: false,
        });
    }

    let mut condition_anchor = condition_block.clone();
    condition_anchor.statements.clear();
    body_blocks.push(condition_anchor.clone());

    let mut body_container = IrBlockContainer {
        id: state.next_container_id(),
        blocks: body_blocks,
        kind: IrContainerKind::Block,
        should_emit_exit_label: container.should_emit_exit_label,
    };
    remove_trailing_continue_jump(&mut body_container, condition_anchor.start_address);

    IrStatement::HighLevelDoWhile(IrHighLevelDoWhile {
        condition: if_stmt.condition.clone(),
        body: body_container,
        continue_target: condition_anchor.start_address,
        break_target: container.id,
    })
}

fn as_jump_target_addr(statement: &IrStatement) -> Option<u32> {
    match statement {
        IrStatement::Jump(IrJump { target_address }) => Some(*target_address),
        IrStatement::Block(block) if block.statements.len() == 1 => {
            match block.statements.first() {
                Some(IrStatement::Jump(IrJump { target_address })) => Some(*target_address),
                _ => None,
            }
        }
        _ => None,
    }
}

fn is_leave_to_container(statement: &IrStatement, container_id: u32) -> bool {
    matches!(
        statement,
        IrStatement::Leave(IrLeave { target_container_id }) if *target_container_id == container_id
    )
}

fn is_jump_to_entry(statement: &IrStatement, entry_address: u32) -> bool {
    as_jump_target_addr(statement).is_some_and(|target| target == entry_address)
}

fn remove_trailing_continue_jump(body: &mut IrBlockContainer, continue_target: u32) {
    for block in body.blocks.iter_mut().rev() {
        if block.statements.is_empty() {
            continue;
        }
        let remove = matches!(
            block.statements.last(),
            Some(IrStatement::Jump(IrJump { target_address })) if *target_address == continue_target
        );
        if remove {
            block.statements.pop();
        }
        break;
    }
}

fn max_container_id(container: &IrBlockContainer) -> u32 {
    let mut max_id = container.id;
    for block in &container.blocks {
        for statement in &block.statements {
            max_id = max_id.max(max_container_id_in_statement(statement));
        }
    }
    max_id
}

fn max_container_id_in_statement(statement: &IrStatement) -> u32 {
    match statement {
        IrStatement::Block(block) => block
            .statements
            .iter()
            .map(max_container_id_in_statement)
            .max()
            .unwrap_or(0),
        IrStatement::BlockContainer(container) => max_container_id(container),
        IrStatement::If(if_stmt) => {
            let mut max_id = max_container_id_in_statement(if_stmt.true_statement.as_ref());
            if let Some(false_statement) = if_stmt.false_statement.as_ref() {
                max_id = max_id.max(max_container_id_in_statement(false_statement.as_ref()));
            }
            max_id
        }
        IrStatement::HighLevelSwitch(switch_stmt) => switch_stmt
            .sections
            .iter()
            .flat_map(|section| section.body.statements.iter())
            .map(max_container_id_in_statement)
            .max()
            .unwrap_or(0),
        IrStatement::HighLevelWhile(while_stmt) => max_container_id(&while_stmt.body),
        IrStatement::HighLevelDoWhile(do_while_stmt) => max_container_id(&do_while_stmt.body),
        IrStatement::Assignment(_)
        | IrStatement::Expression(_)
        | IrStatement::VariableDeclaration(_)
        | IrStatement::Jump(_)
        | IrStatement::Leave(_)
        | IrStatement::Return(_)
        | IrStatement::Switch(_) => 0,
    }
}
